### PR TITLE
Cross-chain Asset Config

### PR DESCRIPTION
**Pull request by Dinonard**
**Pull Request Summary**

Adds pallet to handle mapping between local asset and cross-chain asset.
Also handles payment rate for foreign assets.

The main idea is to provide simple pallet without intertwining it with `pallet-assets` or reinventing that pallet.

Note that at the moment it's not possible to add genesis config because `MultiLocation` isn't serializable.
We should change this in `polkadot` repo.

**TODO**
- [x] team review
- [x] tests
- [x] benchmarks
- [x] callback for EVM code injection handling

**Check list**
- [ ] contains breaking changes
- [x] adds new feature
- [ ] modifies existing feature (bug fix or improvements)
- [ ] relies on other tasks
- [ ] documentation changes
- [x] tests and/or benchmarks are included
- [ ] changed API client type definition or chain metadata

### CODE REVIEW

File: frame/xc-asset-config/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    pub type AssetIdToLocation<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AssetId, VersionedMultiLocation>;
+
+    /// Mapping from an asset type to an asset id.
+    /// Can be used when receiving a multilocation XCM message to retrieve
+    /// the corresponding asset in which tokens should me minted.
+    #[pallet::storage]
+    #[pallet::getter(fn asset_location_to_id)]
+    pub type AssetLocationToId<T: Config> =
+        StorageMap<_, Blake2_128Concat, VersionedMultiLocation, T::AssetId>;
```
Comment (Dinonard) [inline frame/xc-asset-config/src/lib.rs]: TODO: Change to more efficient twox hashing.

**Comment by Dinonard** [comment]
@akru @Maar-io @codespool please review/approve.

This has been integrated into Shibuya (locally) and tested extensively.

File: frame/xc-asset-config/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+    use parity_scale_codec::HasCompact;
+    use sp_std::boxed::Box;
+    use xcm::{v1::MultiLocation, VersionedMultiLocation};
+
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    /// Callback definition trait for cross-chain asset registration/deregistration notifications.
+    pub trait XcAssetChangedCallback<T: Config> {
```
Comment (akru) [inline frame/xc-asset-config/src/lib.rs]: I believe we could easily drop Callback from name without loss in means.

File: frame/xc-asset-config/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        /// Will be called by pallet when new asset Id has been registered
+        fn xc_asset_registered(asset_id: T::AssetId);
+
+        /// Will be called by pallet when asset Id has been unregistered
+        fn xc_asset_unregistered(asset_id: T::AssetId);
+    }
+
+    /// Defines conversion between asset Id and asset type
+    pub trait AssetLocationGetter<AssetId> {
```
Comment (akru) [inline frame/xc-asset-config/src/lib.rs]: Why don’t keep same approach like XcAssetLocation and xc_ prefix for methods?

File: frame/xc-asset-config/src/lib.rs (hunk 2 above)
Comment (Dinonard) [reply frame/xc-asset-config/src/lib.rs]: Sure.

File: frame/xc-asset-config/src/lib.rs (hunk 3 above)
Comment (Dinonard) [reply frame/xc-asset-config/src/lib.rs]: Sure, I can change it.

**Review by akru — APPROVED** [review]
LGatM

**Pull request merged**